src/server/pfs/server/source.go

```diff
@@ -119,38 +44,50 @@ func (s *source) Iterate(ctx context.Context, cb func(*pfs.FileInfo, fileset.Fil
 	fs1 := s.getReader()
 	fs2 := s.getReader()
 	s2 := newStream(ctx, fs2)
-	cache := make(map[string][]byte)
+	cache := make(map[string]*pfs.FileInfo)
 	return fs1.Iterate(ctx, func(fr fileset.File) error {
 		idx := fr.Index()
 		fi := &pfs.FileInfo{
 			File: client.NewFile(s.commit.Repo.Name, s.commit.ID, idx.Path),
 		}
-		if s.computeHashes {
-			var err error
-			var hashBytes []byte
-			if indexIsDir(idx) {
-				hashBytes, err = computeHash(cache, s2, idx.Path)
+		if s.full {
+			cachedFi, ok := checkFileInfoCache(cache, idx)
+			if ok {
+				fi.FileType = cachedFi.FileType
+				fi.SizeBytes = cachedFi.SizeBytes
+				fi.Hash = cachedFi.Hash
+			} else {
+				computedFi, err := computeFileInfo(cache, s2, idx.Path)
 				if err != nil {
 					return err
 				}
-			} else {
-				hashBytes = computeFileHash(idx)
+				fi.FileType = computedFi.FileType
+				fi.SizeBytes = computedFi.SizeBytes
+				fi.Hash = computedFi.Hash
 			}
-			fi.Hash = hashBytes
-		}
-		if err := cb(fi, fr); err != nil {
-			return err
 		}
-		delete(cache, idx.Path)
```

Comment:
Doesn't lexicographical order ensure that a parent is emitted before it's child?  By the time anything is emitted it's FileInfo is no longer needed by any path coming later.
Why can't we delete here?

Reply:
Yes, but for the subsequent child files we need to check if the parent directory info is cached. This allows us to correctly avoid progressing the second stream when it has already been progressed past a file. When we compute a directory info we progress the second stream past all of the files / directories in that directory.

Reply:
We can delete a directory when we are past all of its files / directories, I just need to put more thought into how we handle that. I'm not too concerned about this caching for now, which is why I left it as a TODO.